### PR TITLE
Fix public link path for persistent locks

### LINKED ISSUE #34225: Parent share link path leaked through lock info
### Steps

1. Create a folder "/abc/test/a"
1. Lock "/abc" with exclusive lock
1. Share "/abc/test" with link
1. Query lock info with PROPFIND on public webdav from link:
```
curl -u $linktoken:\
  -X PROPFIND \
  --data-binary "@propfind.xml" \
  -H "Content-Type: text/xml" \
'http://localhost/owncloud/public.php/webdav/x' | xmllint --format -
```

### Expected result
Lock root is always "public.php/webdav/"

### Actual result
Lock root contains the parent folder name: "public.php/webdav/abc" which leaks the parent folder names and itself isn't even a valid folder since there is no "abc" in "test".

### Versions
Observed on master while fixing https://github.com/owncloud/core/pull/34224

Making that p2 as this would introduce a security issue about information disclosure.

@DeepDiver1975 @davitol FYI

### PR DESCRIPTION
**Pull request by PVince81**
## Description
When querying lock info on public links, always return a path relative
to the public webdav endpoint.

## Related Issue
Fixes https://github.com/owncloud/core/issues/34225

## How Has This Been Tested?
- [x] Unit tests
With "/public/share" being the link share:
- [x] TEST: set lock on "/public", query locks on "public.php/webdav/sub" returns lock root "public.php/webdav/"
- [x] TEST: set lock on "/public/share", query locks on "public.php/webdav/sub" returns lock root "public.php/webdav/"
- [x] TEST: set lock on "/public/share/sub", query locks on "public.php/webdav/sub" returns lock root "public.php/webdav/sub"
- [x] TEST: set lock on "/public", set another lock on "public.php/webdav/sub" returns empty lock root ""
- [x] TEST: set lock on "/public/share/sub", set another lock on "public.php/webdav/sub" returns lock root "/sub"

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:

- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

### CODE REVIEW

**Comment by PVince81** [comment]
@davitol FYI see test plan in the section above

@DeepDiver1975 please review

**Comment by davitol** [comment]
Spotted by @individual-it:

with a PROPFIND a client can find out the locktoken, so a nasty client can overwrite resources even they are locked by an other user.

**Steps to reproduce**
1. - User shares a public link and locks a folder
2.- The public does a PROPFIND and gets tould the locktoken
3. - Using that token the public can overwrite locked files

**Comment by PVince81** [comment]
Idea: only expose the lock token if the lock owner is NULL, which refers to a lock that has been created through the public API. if the lock was created by a logged in user, don't reveal it.

@DeepDiver1975 thoughts ?

**Comment by PVince81** [comment]
ok I think I'll always hide the lock token for public users. there is no UI to unlock there and whoever created the lock already has the token somewhere

the file owner can still unlock with the UI as they have the token

**Comment by PVince81** [comment]
note: also tested trying to lock "sub" through public webdav while "public" was locked.
the error response also contains the correct non-leaked lock root:
```
curl -u 6OVGtYzOTEmozQB: -X LOCK http://localhost/owncloud/public.php/webdav/sub -d "<?xml version='1.0' encoding='UTF-8'?><d:lockinfo xmlns:d='DAV:'> <d:lockscope><d:exclusive/></d:lockscope></d:lockinfo>" -H 'Depth: infinite'
<?xml version="1.0" encoding="utf-8"?>
<d:error xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns">
  <s:exception>Sabre\DAV\Exception\ConflictingLock</s:exception>
  <s:message/>
  <d:no-conflicting-lock>
    <d:href/>
  </d:no-conflicting-lock>
</d:error>
```

**Comment by PVince81** [comment]
- [ ] not sure if bug: lock root as returned in lockdiscovery contains full path with public.php while lock root in conflict exception response contains a relative path

**Comment by PVince81** [comment]
https://tools.ietf.org/html/rfc4918#section-6.5 says that

> Servers MAY make lock tokens publicly readable (e.g., in the DAV:
   lockdiscovery property). 

which means we should be able to hide them.

However Sabre still returns the lock even when empty... Will need an upstream patch.

**Comment by PVince81** [comment]
Upstream PR: https://github.com/sabre-io/dav/pull/1130

**Comment by individual-it** [comment]
@ownclouders rebase

**Comment by ownclouders** [comment]
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

**Review by DeepDiver1975 — APPROVED** [review]

**Comment by individual-it** [comment]
pushed a couple of acceptance tests
@PVince81 @phil-davis @davitol  please review test code and expectations

File: tests/acceptance/features/apiWebdavLocks/publicLink.feature [hunk 1]
```diff
@@ -66,3 +66,188 @@ Feature: persistent-locking in case of a public link
       | old      | exclusive  |
       | new      | shared     |
       | new      | exclusive  |
+
+  Scenario: public tries to lock a folder inside an exclusivly locked folder
```
Comment (phil-davis) [inline tests/acceptance/features/apiWebdavLocks/publicLink.feature]: @individual-it s/exclusivly/exclusively

File: tests/acceptance/features/apiWebdavLocks/publicLink.feature [hunk 2]
```diff
@@ -66,3 +66,188 @@ Feature: persistent-locking in case of a public link
...
+        And user "user0" has created a public link share of folder "PARENT" with change permission
+    And user "user0" has locked folder "PARENT" setting following properties
```
Comment (phil-davis) [inline tests/acceptance/features/apiWebdavLocks/publicLink.feature]: different indent starting from here @individual-it 
(maybe it is the indent above that is wrong, and correct indent starts here)

File: tests/acceptance/features/apiWebdavLocks/publicLink.feature [hunk 3]
```diff
@@ -66,3 +66,188 @@ Feature: persistent-locking in case of a public link
...
+      | lock-scope |
+      | shared     |
+      | exclusive  |
+
+  Scenario Outline: lockdiscovery file in a subfolder of a public link when subfolder is locked
+    Given user "user0" has created a public link share of folder "PARENT" with change permission
+    And user "user0" has locked folder "PARENT/CHILD" setting following properties
+      | lockscope | <lock-scope> |
+    When the public gets the following properties of entry "/CHILD/child.txt" in the last created public link using the WebDAV API
+     | d:lockdiscovery |
```
Comment (phil-davis) [inline tests/acceptance/features/apiWebdavLocks/publicLink.feature]: 1 space of indent missing from here - happens in the following scenarios also

**Review by phil-davis — COMMENTED** [review]
@individual-it  acceptance tests look good, just little comments

**Review by phil-davis — APPROVED** [review]

**Comment by PVince81** [comment]
stable10: https://github.com/owncloud/core/pull/34267

I forgot to push the solution for https://github.com/owncloud/core/pull/34229#issuecomment-457225278, it will be in a separate PR

**Pull request merged**